### DISCUSSION TITLE
Detecting columns used in DISTINCT

### DISCUSSION THREAD

**Discussion by sampsonsl** [JSQLParser/JSqlParser#2014] [General]
New to JSQLParser.  I'd like to use it to identify columns used with specific SQL operations, like DISTINCT. 
Suggestion on how this is best done, or pointers to examples appreciated. Thanks.

**Comment by manticore-projects** [comment]
Greetings. 

You would need to provide at least some sample SQLs to illustrate your question please.

**Reply by sampsonsl** [reply]
Thanks @manticore-projects

For example consider the following two SQL queries, both of which have a DISTINCT clause.

For the first query I want to detect that column s_i_id is used in the DISTINCT clause.
For the second query I want to detect that columns c_w_id, c_d_id are used in a DISTINCT clause.

The solution should also handle cases where a single query has multiple DISTINCT clauses, or none at all.

Thank you very much for any pointers and suggestions!

---

SELECT DISTINCT s_i_id
FROM stock
WHERE s_w_id = 1 AND s_quantity < 40;

SELECT COUNT(*)
FROM (SELECT DISTINCT c_w_id, c_d_id
FROM customer) AS distinct_combinations;

**Reply by manticore-projects** [reply]
Greetings.

You will need to traverse the AST and then for any `PlainSelect` check the `Distinct` property. For the 2nd example, [your AST looks like this](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MoUQMiDCAqAEkHkCqA5aAKAVASgFADEAlBAWVnVAhlgBEBJYaOlagYwH0B3dgSwBMANLA59efAsTKsArgGcALgHsAtgFMATtlgBBYLD48FPAHat57VioBGJgIbyei47IDcQA):

```text
SQL Text
 └─Statements: statement.select.PlainSelect
    ├─selectItems: statement.select.SelectItem
    │  └─expression: expression.Function
    │     └─ExpressionList: *
    └─fromItem: statement.select.ParenthesedSelect
       ├─Alias:  AS distinct_combinations
       └─select: statement.select.PlainSelect
          ├─Distinct: DISTINCT
          ├─selectItems: statement.select.SelectItem
          │  └─Column: c_w_id
          ├─selectItems: statement.select.SelectItem
          │  └─Column: c_d_id
          └─Table: customer
```

Parsing and traversing the AST is explained here: https://manticore-projects.com/JSQLParser/usage.html#parse-a-sql-statement

**Reply by sampsonsl** [reply]
Thanks. What is the specific method used to check the Distinct property?

**Reply by manticore-projects** [reply]
See JAVA API: https://manticore-projects.com/JSQLParser/javadoc_snapshot.html#plainselect
Use the method `getDistinct()` to access the `Distinct` object.

**Reply by sampsonsl** [reply]
Thank you for the guidance!

**Reply by sampsonsl** [reply]
Returning to this, using getDistinct() appears to work for cases where DISTINCT is used directly in  the SELECT like my two earlier examples. However, "COUNT( DISTINCT <column>)"  produces getDistinct() == null.  

For example, a query like this: 

SELECT COUNT(DISTINCT s_i_id) FROM stock; 

In this case, I think the parser is including s_i_d under COUNT as a function.   I'm not clear how to find and associate it with DISTINCT.   Suggestions appreciated. Thanks!

**Reply by manticore-projects** [reply]
It's a complete different type of **distinct**, belonging to a function call:

```text
SQL Text
 └─Statements: statement.select.PlainSelect
    ├─selectItems: statement.select.SelectItem
    │  └─expression: expression.Function
    │     └─ExpressionList: s_i_id
    └─Table: stock
```

Java API: http://217.160.215.75:8080/jsqlformatter/JSQLParser/javadoc_snapshot.html#function
Which again has the method: `isDistinct()`.
